lib/vtls/vtls_scache.c

```diff
@@ -0,0 +1,874 @@
+    r = Curl_dyn_add(&buf, ":QUIC");
+    break;
+  case TRNSPRT_UNIX:
+    r = Curl_dyn_add(&buf, ":UNIX");
+    break;
+  default:
+    r = Curl_dyn_addf(&buf, ":TRNSPRT-%d", peer->transport);
+    break;
+  }
+  if(r)
+    goto out;
+
+  if(!ssl->verifypeer) {
+    r = Curl_dyn_add(&buf, ":NO-VRFY-PEER");
+    if(r)
+      goto out;
+  }
+  if(!ssl->verifyhost) {
+    r = Curl_dyn_add(&buf, ":NO-VRFY-HOST");
+    if(r)
+      goto out;
+  }
+  if(ssl->verifystatus) {
+    r = Curl_dyn_add(&buf, ":VRFY-STATUS");
+    if(r)
+      goto out;
+  }
+  if(!ssl->verifypeer || !ssl->verifyhost) {
+    if(cf->conn->bits.conn_to_host) {
+      r = Curl_dyn_addf(&buf, ":CHOST-%s", cf->conn->conn_to_host.name);
+      if(r)
+        goto out;
+    }
+    if(cf->conn->bits.conn_to_port) {
+      r = Curl_dyn_addf(&buf, ":CPORT-%d", cf->conn->conn_to_port);
+      if(r)
+        goto out;
+    }
+  }
+
+  if(ssl->version || ssl->version_max) {
+    r = Curl_dyn_addf(&buf, ":TLSVER-%d-%d", ssl->version,
+                      (ssl->version_max >> 16));
+    if(r)
+      goto out;
+  }
+  if(ssl->ssl_options) {
+    r = Curl_dyn_addf(&buf, ":TLSOPT-%x", ssl->ssl_options);
+    if(r)
+      goto out;
+  }
+  if(ssl->cipher_list) {
+    r = Curl_dyn_addf(&buf, ":CIPHER-%s", ssl->cipher_list);
+    if(r)
+      goto out;
+  }
+  if(ssl->cipher_list13) {
+    r = Curl_dyn_addf(&buf, ":CIPHER13-%s", ssl->cipher_list13);
+    if(r)
+      goto out;
+  }
+  if(ssl->curves) {
+    r = Curl_dyn_addf(&buf, ":CURVES-%s", ssl->curves);
+    if(r)
+      goto out;
+  }
+  if(ssl->verifypeer) {
+    r = cf_ssl_peer_key_add_path(&buf, "CA", ssl->CAfile);
+    if(r)
+      goto out;
+    r = cf_ssl_peer_key_add_path(&buf, "CApath", ssl->CApath);
+    if(r)
+      goto out;
+    r = cf_ssl_peer_key_add_path(&buf, "CRL", ssl->CRLfile);
+    if(r)
+      goto out;
+    r = cf_ssl_peer_key_add_path(&buf, "Issuer", ssl->issuercert);
+    if(r)
+      goto out;
+    if(ssl->cert_blob) {
+      r = cf_ssl_peer_key_add_hash(&buf, "CertBlob", ssl->cert_blob);
+      if(r)
+        goto out;
+    }
+    if(ssl->ca_info_blob) {
+      r = cf_ssl_peer_key_add_hash(&buf, "CAInfoBlob", ssl->ca_info_blob);
+      if(r)
+        goto out;
+    }
+    if(ssl->issuercert_blob) {
+      r = cf_ssl_peer_key_add_hash(&buf, "IssuerBlob", ssl->issuercert_blob);
+      if(r)
+        goto out;
+    }
+  }
+  if(ssl->pinned_key && ssl->pinned_key[0]) {
+    r = Curl_dyn_addf(&buf, ":Pinned-%s", ssl->pinned_key);
+    if(r)
+      goto out;
+  }
+
+  if(ssl->clientcert && ssl->clientcert[0]) {
+    r = Curl_dyn_add(&buf, ":CCERT");
+    if(r)
+      goto out;
+  }
+#ifdef USE_TLS_SRP
+  if(ssl->username || ssl->password) {
+    r = Curl_dyn_add(&buf, ":SRP-AUTH");
+    if(r)
+      goto out;
+  }
+#endif
+
+  if(!tls_id || !tls_id[0]) {
+    r = CURLE_FAILED_INIT;
+    goto out;
+  }
+  r = Curl_dyn_addf(&buf, ":IMPL-%s", tls_id);
+  if(r)
+    goto out;
+
+  *ppeer_key = Curl_dyn_strdup(&buf);
```

Comment:
There's no need to strdup this buffer here, you can just get the pointer from the buffer and use that and skip freeing the dynbuf thus saving the extra malloc/free:

~~~c
*ppeer_key = Curl_dyn_ptr(&buf);
/* no freeing of the dynbuf, returning the pointer */
~~~

Reply:
Since the `peer_key` sticks around for some time, I'd thought saving the extra bytes that dynbuf might have grown to is worth it. Debatable, sure.

Reply:
Ah, well that's not a bad thought but I think that's still second-guessing the underlying solution a little too much. What made me find this instance in the first place is how this new `Curl_dyn_strdup()` does not duplicate the full buffer, which actually may contain null bytes, which makes this function stand out a little among the other dynbuf functions.

I got me thinking about if it would be worth either changing the function for this, document the exception clearly or perhaps just not use it. Then I noticed that it could actually be removed quite easily, which would then also save the *it is an exception*-dilemma...

Reply:
Fair enough. Then I'll remove it.

Reply:
Removed `Curl_dyn_strdup()`  again and added `Curl_dyn_take()`. This may sound superfluous, since the dynbuf goes out of scope anyway, but I feel better with a clear transfer of ownership.

Reply:
I think that's a good idea. Once this gets merged, we should probably consider making use of this more widely for similar use cases.

Reply:
Btw, document the newcomer it in `DYNBUF.md` as well!

Reply:
Done.